### ISSUE TITLE
TestContext.Progress.Write writes new line

### ISSUE THREAD

**Issue by CharliePoole** [nunit/nunit-console#72]
@yoyos06 commented on [Mon Oct 03 2016](https://github.com/nunit/nunit/issues/1819)

Hi,

The TestContext.Progress.Write method now works exactly the same as TestContext.Progress.WriteLine because it writes a newline character at the end.

---

@rprouse commented on [Mon Oct 03 2016](https://github.com/nunit/nunit/issues/1819#issuecomment-251127409)

Thanks for the report, we will look into it.

---

@CharliePoole commented on [Mon Oct 03 2016](https://github.com/nunit/nunit/issues/1819#issuecomment-251230619)

@yoyos06 Actually, TestContext.Progress.Write doesn't write a newline after the text. It's done in NUnit console.

That might seem like an academic distinction, but it's actually important. The framework puts out events which are received and handled by runners. Each runner may handle them in a different way. For example, some gui might have a special text box that received all the "Progress" output.

In the case of the console runner, the output is handled in a line-oriented way. Every write currently gets a newline because of some earlier errors where lines intended to be separate were run together. We need to be a little smarter about this.

What needs to happen (I'm writing this for whomever implements the fix) is that no newline should be written until it's actually needed. That means the console's TestEventHandler class has to maintain a bit of state regarding whether the last bit of output had a newline attached. When output arrives from some other test it should supply the newline, if it wasn't there already, before writing the new output. It's potentially complicated, but doable.

Since this is a console issue, I'm copying this there. I'm leaving this issue open to deal with NUnitLite, which no doubt has the same problem.

**Comment by rvignesh89** [comment]
@ChrisMaddock @CharliePoole I'd like to help out with this issue.

Based on what I saw removing the following condition in TestEventHandler can help to fix this problem. 

```
private void WriteOutputLine(string text, ColorStyle color)
        {
            using (new ColorConsole(color))
            {
                _outWriter.Write(text);

                //// Some labels were being shown on the same line as the previous output
                //if (!text.EndsWith(Environment.NewLine))
                //{
                //    _outWriter.Write("a");
                //    _outWriter.WriteLine();
                //}
            }
        }
```

Please advise if I'm missing something here.

**Comment by CharliePoole** [comment]
@rvignesh89 Please look back at the original description of the problem where I describe what needs to be done. The problem is that single character or partial-line writes need to eventually get a newline. Doing one for every write is too much. Never writing a newline is too little. The TestEventHandler has to keep track of whether a newline is needed to end the latest output and then write that newline whenever the source of the output changes. The WriteOutputLine method is too late to do it, because we already decided when we call it that we want to write a full line with a linefeed. I believe it has to be fixed when handling TestOutput events only and not for every line that is written from whatever source.

**Comment by rvignesh89** [comment]
@CharliePoole this is what I understood from your description above. 

```
            [Test]
            public void DummyTest()
            {
                TestContext.Progress.Write("Progress 1");
                TestContext.Progress.Write("Progress 2");
            }

            [Test]
            public void DummTest2()
            {
                TestContext.Progress.Write("Progress 3");
            }
```

Current output looks like this

```
=> NUnit.Tests.Assemblies.MockTestFixture.DummTest2
Progress 3
=> NUnit.Tests.Assemblies.MockTestFixture.DummyTest
Progress 1
Progress 2
```

Whereas, you'd like them to look like this?

```
=> NUnit.Tests.Assemblies.MockTestFixture.DummTest2
Progress 3
=> NUnit.Tests.Assemblies.MockTestFixture.DummyTest
Progress 1Progress 2
```

**Comment by CharliePoole** [comment]
Yes, unless they are running in parallel, when you might see this:

```
=> NUnit.Tests.Assemblies.MockTestFixture.DummyTest
Progress 1
=> NUnit.Tests.Assemblies.MockTestFixture.DummyTest2
Progress 3
=> NUnit.Tests.Assemblies.MockTestFixture.DummyTest
Progress 2
```

Note: I assume we are using the default `--labels=On` here.

Before displaying the second label, the handler has to detect the fact that there is a pending newline that must be put out. Since immediate messages are... well... immediate, we need to allow for them coming from any test in any order. That's what makes this a non-trivial fix.

It's also because the fix is non-trivial that we are temporarily treating all output as a WriteLine, even if it was just a Write. The extra newline is a lesser evil than mixing output from different tests.

**Comment by MoFtZ** [comment]
Have raised a pull request to **partly** fix this issue. Refer #209

It also requires the corresponding pull request in the NUnit Framework. Refer https://github.com/nunit/nunit/pull/2129

**Comment by y-polonsk** [comment]
Hi,

What version of NUnit was the issue fixed? I m using NUnit 3.14.0 and facing exactly the same issue. It seems that 3.14.0 is the latest from 3.x versions.

Thanks.

**Comment by CharliePoole** [comment]
@y-polonsk The comment by @rprouse indicates it was fixed in 3.7, seven years ago.

Problems that look "exactly the same" as another issue are often only the same on the surface. Best bet this long after the fix would be to create a new issue describing exactly what you are seeing as a current bug. 

The latest release of the console is 3.17.0, so I suggest you try that before reporting a bug.

**Comment by y-polonsk** [comment]
Thanks.  May I ask, I m kind of confused, the console runner seems to be a separate runner for NUnit.  I m running my tests via "dotnet test" and using Microsoft.NET.Test.Sdk adapter version 17.10.0.  I tried removing NUnit and adding NUnit.Console or NUnit.ConsoleRunner to my dependencies, but then it doesn't compile (cannot find NUnit namespace).  Do i need to do something different to use the console runner?

**Comment by CharliePoole** [comment]
The console runner is the original runner for nunit, created about 24 years ago. The version 3 console is called `nunit3-console.exe` and the version 4 being developed is `nunit4-console.exe`. The console runner has to be installed in some way: nuget, chocolatey, msi or zip. 

Your tests reference the nunit framework, because that's where things like Assert are defined. They don't reference the console. In principle, a test isn't supposed to know who is executing it. :-)

In any case, if you are not using the console runner, then you can't be having the same problem as the OP.

**Comment by y-polonsk** [comment]
I see, thanks. But I m using TestContext.Progress.Write() and it prints new line.  In this case, the problem is perhaps with NUnit3TestAdapter? I m using its version 4.5.0, which is the latest version on nuget.

**Comment by CharliePoole** [comment]
All the different text outputs from `nunit.framework` are translated into events, which are sent to whatever program is running the tests. That runner gets to decide whether and how to display the output. In addition, any output to `stdout` is added to the `<output>` element in the test result file.

So, if your runner is the VisualStudio Test Window, then you need to check with the adapter project.
